[AMDPGU] Auto-upgrade ELF mangling in the data layout

jhuber6 opened the pull request:
Summary:
The changes in https://github.com/llvm/llvm-project/pull/163011 caused
all ELF platforms to default to ELF mangling. We want to auto upgrade
this for linking in new programs to old ones.

jhuber6 commented:
> Generally we just update the four .ll files in the device library. Are there other libraries we're concerned about now where a warning wouldn't be desired?

It's all the files with a data layout, and if we changed this I suppose there should be auto-upgrade anyway.

b-sumner commented:
> > Generally we just update the four .ll files in the device library. Are there other libraries we're concerned about now where a warning wouldn't be desired?
> 
> It's all the files with a data layout, and if we changed this I suppose there should be auto-upgrade anyway.

Aren't there only 4 files?  Isn't the warning a good thing in general?

jhuber6 commented:
> > > Generally we just update the four .ll files in the device library. Are there other libraries we're concerned about now where a warning wouldn't be desired?
> > 
> > 
> > It's all the files with a data layout, and if we changed this I suppose there should be auto-upgrade anyway.
> 
> Aren't there only 4 files? Isn't the warning a good thing in general?

OpenMP currently uses these, HIP has `ockl` in addition I believe.
```
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/ocml.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_daz_opt_off.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_unsafe_math_off.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_finite_only_off.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_correctly_rounded_sqrt_on.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_wavefrontsize64_off.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_isa_version_1030.bc"
 "-mlink-builtin-bitcode" "/opt/rocm/amdgcn/bitcode/oclc_abi_version_600.bc"
```

jhuber6 committed:
[AMDPGU] Auto-upgrade ELF mangling in the data layout

Summary:
The changes in https://github.com/llvm/llvm-project/pull/163011 caused
all ELF platforms to default to ELF mangling. We want to auto upgrade
this for linking in new programs to old ones.

contains

jhuber6 commented:
This applies to all the data layouts, I could make it only apply to this very specific one if necessary, but I'm assuming this is what we want.

nikic commented:
All data layout changes are generally expected to come with an auto-upgrade.

shiltian commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -6055,35 +6054,43 @@ std::string llvm::UpgradeDataLayoutString(StringRef DL, StringRef TT) {
+    }
+
+    // Upgrade the ELF mangling mode.
+    if (!DL.starts_with("e-m:"))
```
why not `contains`, like what others are doing, but assert this will be the start? I don't think they are cannocialized here.
